onnxscript/function_libs/torch_aten/ops/core.py

```diff
@@ -269,44 +269,80 @@ def aten_any(self: TensorType) -> TensorType:
     raise NotImplementedError()
 
 
-@torch_op("aten::arange")
+def _range_supported(dtype: int) -> bool:
+    """Returns true if the dtype is supported by the ONNX Range op."""
+    return dtype in {
+        DOUBLE.dtype,
+        FLOAT.dtype,
+        INT16.dtype,
+        INT32.dtype,
+        INT64.dtype,
+    }
+
+
+@torch_op("aten::arange", trace_only=True)
 def aten_arange(end: Union[DOUBLE, FLOAT, INT16, INT32, INT64], dtype: int = -1) -> TensorType:
     # arange(Scalar end, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
 
-    # Cast input to double if dtype is specified, because the input dtype may be e.g. bool
-    # which Range does not support. The output type is ensured because the output
-    # is casted to the specified dtype.
-    if dtype != -1:
-        end = op.Cast(end, to=DOUBLE.dtype)
+    # NOTE: trace_only because both if branches need to be the same type, but we have
+    # a cast in the if branch.
 
-    result = op.Range(0, end, 1)
-    if dtype != -1:
-        result = op.Cast(result, to=dtype)
+    if _range_supported(dtype):
+        end = op.Cast(end, to=dtype)
+        zero = op.Cast(0, to=dtype)
+        one = op.Cast(1, to=dtype)
+        result = op.Range(zero, end, one)
+    else:
+        # Cast input to float if dtype is not supported by Range,
+        # because the input dtype may be e.g. bool
```

Comment:
Is this used, eg., for 16-bit floats? Just wondering, since bool seems like a very special odd-case, though I guess a user could use it.

Reply:
Hopefully not for complex, which is tricky

Reply:
Not complex (and not bool actually):

![image](https://user-images.githubusercontent.com/11205048/217684440-7c3fa99d-cde6-4fb7-bf42-c8cadf9c5b18.png)

Reply:
uint8, int8, float16, bfloat16